### PR TITLE
Implement filter pushdown for PostgreSQL

### LINKED ISSUE #3414: Implement filter pushdown
### What should be done?

We should implement collection query filter pushdown as in `pgdb`.

Relevant integration tests should pass.

### Where?

https://github.com/FerretDB/FerretDB/blob/main/internal/backends/postgresql/collection.go

### Definition of Done

- unit tests added/updated;
- integration/compatibility tests added/updated;
- spot refactorings done.

### PR DESCRIPTION
**Pull request by noisersup**
# Description

Closes #3414.

## Readiness checklist

- [ ] I added/updated unit tests (and they pass).
- [ ] I added/updated integration/compatibility tests (and they pass).
- [ ] I added/updated comments and checked rendering.
- [x] I made spot refactorings.
- [ ] I updated user documentation.
- [ ] I ran `task all`, and it passed.
- [x] I ensured that PR title is good enough for the changelog.
- [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Milestone (`Next`), Labels, Project and project's Sprint fields.
- [x] I marked all done items in this checklist.

### CODE REVIEW

**Review by AlekSi — COMMENTED** [review]
There is a merge conflict now

File: internal/backends/postgresql/collection.go [hunk 1]
```diff
@@ -73,14 +77,22 @@ func (c *collection) Query(ctx context.Context, params *backends.QueryParams) (*
 		}, nil
 	}
 
-	// TODO https://github.com/FerretDB/FerretDB/issues/3414
 	q := fmt.Sprintf(
 		`SELECT %s FROM %s`,
```
Comment (AlekSi) [inline internal/backends/postgresql/collection.go:81]: Let's extract that part into a common function too, because we will change it soon

File: internal/backends/postgresql/collection.go [hunk 2]
```diff
@@ -261,20 +275,29 @@ func (c *collection) Explain(ctx context.Context, params *backends.ExplainParams
...
-		return &backends.ExplainResult{
-			QueryPlanner: must.NotFail(types.NewDocument()),
-		}, nil
+		res.QueryPlanner = must.NotFail(types.NewDocument())
+		return res, nil
 	}
 
-	// TODO https://github.com/FerretDB/FerretDB/issues/3414
 	q := fmt.Sprintf(
 		`EXPLAIN (VERBOSE true, FORMAT JSON) SELECT %s FROM %s`,
```
Comment (AlekSi) [inline internal/backends/postgresql/collection.go:283]:
```go
"EXPLAIN (VERBOSE true, FORMAT JSON) " + commonQueryPartReturnedByANewFunction
```

File: internal/backends/postgresql/query.go [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+func prepareSelectClause(db, table string) string {
+	return fmt.Sprintf(
+		`SELECT %s FROM %s`,
+		metadata.DefaultColumn,
+		pgx.Identifier{db, table}.Sanitize(),
+	)
+}
+
+// prepareWhereClause adds WHERE clause with given filters to the query and returns the query and arguments.
+func prepareWhereClause(p *Placeholder, sqlFilters *types.Document) (string, []any, error) {
```
Comment (chilagrow) [inline internal/backends/postgresql/query.go:53]: Do we not want to add tests for this?

**Review by chilagrow — COMMENTED** [review]
Just question about test

**Review by AlekSi — COMMENTED** [review]
`task fmt lint`

File: internal/backends/postgresql/query.go [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+	"github.com/FerretDB/FerretDB/internal/backends/postgresql/metadata"
+	"github.com/FerretDB/FerretDB/internal/handlers/sjson"
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/iterator"
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
+	"github.com/FerretDB/FerretDB/internal/util/must"
+)
+
+// Placeholder stores the number of the relevant placeholder of the query.
+type Placeholder int
```
Comment (AlekSi) [inline internal/backends/postgresql/query.go:35]: Let's use `metadata.Placeholder` instead: https://github.com/FerretDB/FerretDB/blob/main/internal/backends/postgresql/metadata/placeholder.go

**Review by AlekSi — COMMENTED** [review]
https://github.com/FerretDB/FerretDB/pull/3482#discussion_r1347628520

File: internal/backends/postgresql/query_test.go [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+	// WHERE clauses occurring frequently in tests
+	whereContain := " WHERE _jsonb->$1 @> $2"
+	whereGt := " WHERE _jsonb->$1 > $2"
+	whereNotEq := ` WHERE NOT ( _jsonb ? $1 AND _jsonb->$1 @> $2 AND _jsonb->'$s'->'p'->$1->'t' = `
+
+	for name, tc := range map[string]struct {
+		filter   *types.Document
+		expected string
+		skip     string
+		args     []any // if empty, check is disabled
```
Comment (chilagrow) [inline internal/backends/postgresql/query_test.go:43]: Why do we disable check for args? I know we did it this way previously but just commenting it makes it difficult for readers.

Currently, test cases are a bit difficult to read because expected where clause such as `whereContain` is extracted top of the function, so each test case I go up to the top of the test to read what is expected.

Seeing something like below to me is easier to tell input `filter`, `expected` where clause and `args`. :hugs: 
```
		"EqObjectID": {
			filter: must.NotFail(types.NewDocument(
				"v", must.NotFail(types.NewDocument("$eq", objectID)),
			)),
			expected: " WHERE _jsonb->$1 @> $2",
			args: []any{`v`, `"6256c5baadc0ffeeffffff"`},
		},
```

Comment (chilagrow) [reply internal/backends/postgresql/query_test.go:43]: It's an optional comment, because after all we test what we want to test.

Comment (noisersup) [reply internal/backends/postgresql/query_test.go:43]: We use args mainly to validate that we send proper arguments in edge case scenarios. It's optional as adding expected values such as 42 or "foo" for every test doesn't make a lot of sense, but it takes more lines of code, and creating new tests is just more difficult.

File: internal/backends/postgresql/query.go [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+		switch {
+		case v > maxSafeDouble:
+			sql = `_jsonb->%[1]s > %[2]s`
+			v = maxSafeDouble
+
+		case v < -maxSafeDouble:
+			sql = `_jsonb->%[1]s < %[2]s`
+			v = -maxSafeDouble
+		default:
+			// don't change the default eq query
```
Comment (rumyantseva) [inline internal/backends/postgresql/query.go:217]: Do we need this `default` branch at all?

File: internal/backends/postgresql/query.go [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+		switch {
+		case v > types.MaxSafeDouble:
+			sql = `_jsonb->%[1]s > %[2]s`
+			v = types.MaxSafeDouble
+
+		case v < -types.MaxSafeDouble:
+			sql = `_jsonb->%[1]s < %[2]s`
+			v = -types.MaxSafeDouble
+		default:
+			// don't change the default eq query
```
Comment (rumyantseva) [inline internal/backends/postgresql/query.go:188]: Do we need this `default` branch at all?

**Review by rumyantseva — APPROVED** [review]
LGTM

File: internal/backends/postgresql/query.go (hunk 6 above)
Comment (noisersup) [reply internal/backends/postgresql/query.go:217]: It's not necessary, but at least to me it's easier to understand what do we do in default cases

**Review by AlekSi — APPROVED** [review]

File: internal/backends/postgresql/query_test.go (hunk 5 above)
Comment (chilagrow) [reply internal/backends/postgresql/query_test.go:43]: We always copy and paste for test cases, so I don't think creating new test would be harder :see_no_evil:
